### ISSUE TITLE
RFE: Add "query by upstream ecosystem" REST API

### ISSUE THREAD

**Issue by ncoghlan** [fedora-infra/anitya#326]
The current project details API is accessed based on Anitya's internally allocated numeric ID for a project: https://release-monitoring.org/api/#Retrieve_a_specific_project

This RFE proposes adding an upstream counterpart to the existing distros endpoint: https://release-monitoring.org/api/#Retrieve_a_package_for_a_distro

The minimalist-but-confusing way to do this would be to model upstream ecosystems as equivalent to downstream distros - they'd just be another entry in the distros list. While this would minimise design & implementation effort, I don't think it's the right answer, as it wouldn't provide a clear distinction between publication platforms (PyPI, RubyGems, npm, etc) and integration platforms (Fedora, RHEL, conda, Debian, Nix, etc), and would thus require special casing of the analysis tools.

The other way I see of tackling this without requiring a significant refactoring of the data model is to flag a subset of Backend implementations as actually corresponding to upstream ecosystems by adding a new "Ecosystem" table. Initially this table would just contain the ecosystem name and a reference to the relevant backend implementation, but over time may also gain additional ecosystem specific details like custom version ordering algorithms that capture the details of things like Python's PEP 440.

The key distinction between a mere backend and an Ecosystem is that for a given ecosystem, project names within that ecosystem are required to be unique. There can be multiple "pip" projects within Anitya as a whole, for example, but only one within the PyPI ecosystem and one within the CPAN ecosystem. The "custom" backend, for example, wouldn't define an ecosystem, as you can have projects with the same name hosted on diferent sites.

Given such a mapping from Ecosystem names to backends, and hence to projects using those backends, an ecosystem + name query API can be defined. Since "/api/project/<distro>/<name>" is already claimed, I suggest the following approach:

* add `/api/project/by_ecosystem/<ecosystem>/<name>` as a new lookup endpoint
* add `/api/project/by_distro/<distro>/<name>` as an alias for the existing by distro lookup

I'll start working on a PR for the `/api/project/by_ecosystem/<ecosystem>/<name>` idea, so we can get an initial idea of what this might look like in practice.

**Comment by ncoghlan** [comment]
Draft PR to start adding the Ecosystem model: https://github.com/fedora-infra/anitya/pull/327

It's a bit tough to read at the moment, since it's based on top of the #322 changes, but https://github.com/fedora-infra/anitya/pull/327/commits/ae8f50d7d03d0297f0bbbe6d3d0f5ae47bb474f8 contains the current set of changes specifically related to this issue.

**Comment by ncoghlan** [comment]
The #327 PR should be feature complete now. Rather than the `/api/project/by_ecosystem` endpoint I suggested above, I moved it up a level and made it a `/api/by_ecosystem/` read-only collection.

This nicely avoids the URL pattern conflict that would otherwise exist between "by_ecosystem" and the current distro names endpoint. I also haven't added the `by_distro` alias at this point, since that's an unrelated change purely to make the endpoint a bit more self-documenting, and not needed to address this RFE.
